src/Adapter/Doctrine/ORMAdapter.php

```diff
@@ -197,8 +205,11 @@ protected function getResults(AdapterQuery $query): \Traversable
             ;
         }
 
-        //!$this->query->getHint(self::HINT_DISTINCT) || isset($this->selectedClasses[$joinedDqlAlias])
-        foreach ($builder->getQuery()->iterate([], $this->hydrationMode) as $result) {
+        $query = $builder->getQuery();
+        $event = new ORMAdapterQueryEvent($query);
+        $this->eventDispatcher->dispatch(ORMAdapterEvents::PRE_QUERY, $event);
```

Comment:
You can access the dispatcher via `$state->getTable()`, that's why the adapter does not need to be aware of the EventDispatcher.